kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+    Add(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Sub(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Mul(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    UnnormalizedLagrangeBasis(i32),
+    Cache(CacheId, Box<FoldingExp<C>>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Degree {
+    Zero,
+    One,
+    Two,
+}
+impl<C> FoldingExp<C> {
+    pub(super) fn degree(&self) -> Degree {
+        use Degree::*;
+        match self {
+            FoldingExp::Constant(_) => Zero,
+            FoldingExp::Cell(col) => match col {
+                FoldingColumn::Witness(_) | FoldingColumn::WitnessExtended(_) => One,
+                _ => Zero,
+            },
+            FoldingExp::Double(e) => e.degree(),
+            FoldingExp::Square(e) => &e.degree() * &e.degree(),
+            FoldingExp::Add(e1, e2) | FoldingExp::Sub(e1, e2) | FoldingExp::Mul(e1, e2) => {
+                e1.degree() + e2.degree()
+            }
+            FoldingExp::UnnormalizedLagrangeBasis(_) => Zero,
+            FoldingExp::Cache(_, e) => e.degree(),
+        }
+    }
+}
+impl std::ops::Add for Degree {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) | (One, One) => One,
+            (_, Two) | (Two, _) => Two,
+        }
+    }
+}
+impl std::ops::Mul for &Degree {
+    type Output = Degree;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) => One,
+            (Zero, Two) | (One, One) | (Two, Zero) => Two,
+            (One, Two) | (Two, One) | (Two, Two) => panic!("degree over 2"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Term<C> {
+    pub exp: FoldingExp<C>,
+    pub sign: bool,
+    pub degree: Degree,
+}
+
+impl<C> Term<C> {
+    fn new(exp: FoldingExp<C>, degree: Degree) -> Self {
+        Self {
+            exp,
+            sign: true,
+            degree,
+        }
+    }
+}
+impl<C> std::ops::Neg for Term<C> {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Term {
+            sign: !self.sign,
+            ..self
+        }
+    }
+}
+///A simplified expression with all terms separated by degree
+pub struct IntegratedFoldingExpr<C> {
+    pub(super) degree_0: Vec<(FoldingExp<C>, bool)>,
+    pub(super) degree_1: Vec<(FoldingExp<C>, bool)>,
+    pub(super) degree_2: Vec<(FoldingExp<C>, bool)>,
+}
+
+impl<C> Default for IntegratedFoldingExpr<C> {
+    fn default() -> Self {
+        Self {
+            degree_0: vec![],
+            degree_1: vec![],
+            degree_2: vec![],
+        }
+    }
+}
+// fn extract_terms<C: Clone, F>(exp: FoldingExp<C>) -> Vec<Term<C>>
+pub fn extract_terms<C: Clone, F>(exp: FoldingExp<C>, terms: &mut Vec<Term<C>>, f: &F)
```

Comment:
Regarding the function style I'm tempted to say that it should not operate on a mutable append-only vec (note that append-only property is not enforced on the type level), but instead return `Vec<Term>` which will be concatenated, but maybe that's not as efficient. 

Another point is that I don't see why you have to pass this `map` argument `f` into the next call instead of mapping elements in the current iteration and calling `extract_terms` recursively on the already-transformed data. I guess also efficiency?

Reply:
Yeah, concatenation is the most aesthetic for me, I just didn't want to allocate so many `Vec`s, but maybe the cost is not significant and it's better to do refactor it that way.
I had some issues with the compiler and `f` which led to the unnatural use of recursion, don't look very deep into it, it's just the first approach I could get to compile.
In the latest versions of Rust you get a slightly more understandable error instead of a crash, I may be able to refactor it later into a more straightforward way.